### PR TITLE
EFS encryption at rest

### PR DESCRIPTION
**Pull request by aliaksei-ivanou**
*Issue #, if available:*

*Description of changes:*
- Implementation of a new EFS resource provider
- Bug fixes and improvement of an existing KMS resource provider
- Extending efs-csi-driver
- Adding the documentation covering new functionality

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/spi/types.ts [hunk 1]
```diff
@@ -80,6 +80,7 @@ export interface ApplicationRepository extends GitRepositoryReference {
 export class ResourceContext {
 
     private readonly resources: Map<string, cdk.IResource> = new Map();
+    removalPolicy: cdk.RemovalPolicy | undefined;
```
Comment (shapirov103) [inline lib/spi/types.ts:83]: Why do we need this here? Where is it used?

**Review by shapirov103 — COMMENTED** [review]
@aliaksei-ivanou thank you for this PR, very thorough. 

As I mentioned elsewhere:

1. tab size: 4 (spaces)
2. autoformat: changed lines only

Added one comment on the spi, I will review the rest after GitHub can load diffs by default.

File: lib/spi/types.ts (hunk 1 above)
Comment (aliaksei-ivanou) [reply lib/spi/types.ts:83]: My apologies, I reversed all the unnecessary changes in the code due to the auto-formatting.

**Comment by aliaksei-ivanou** [comment]
> 

My apologies, I reversed all the unnecessary changes in the code due to the auto-formatting.

File: lib/spi/types.ts (hunk 1 above)
Comment (aliaksei-ivanou) [reply lib/spi/types.ts]: addressed

**Review by shapirov103 — COMMENTED** [review]
Great work, raised the bar compared to the bare minimum that was expected. I added a couple of minor comments. Please take a look, I will kick off testing after it.

File: lib/resource-providers/efs.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    efsFileSystem = efs.FileSystem.fromFileSystemAttributes(
+      context.scope,
+      this.options.name,
+      {
+        securityGroup: securityGroup,
+        fileSystemId: this.options.fileSystemId,
+      }
+    );
+
+    if (!efsFileSystem) {
```
[the remark this answers is not in the document]
Comment (aliaksei-ivanou) [reply lib/resource-providers/efs.ts]: addressed

File: docs/addons/efs-csi-driver.md [hunk 3]
```diff
@@ -35,7 +35,7 @@ const blueprint = blueprints.EksBlueprint.builder()
 - `version`: Version of the EFS CSI Driver add-on to be installed. Version 2.2.3 will be installed by default if a value is not provided
 - `replicaCount`: Number of replicas to be deployed. If not provided, two replicas will be deployed. Note that the number of replicas
   should be less than or equal to the number of nodes in the cluster otherwise some pods will be left of pending state
-
+- `kmsKeys`: List of KMS keys to be used for encryption-at-rest
```
[the remark this answers is not in the document]
Comment (aliaksei-ivanou) [reply docs/addons/efs-csi-driver.md]: addressed

**Review by shapirov103 — APPROVED** [review]
lgtm

**Comment by shapirov103** [comment]
Failed e2e, from what I see it is a failure on `make run-test`:

```
FAIL test/resource-providers/resource-proxy.test.ts
  ● Test suite failed to run
    test/resource-providers/resource-proxy.test.ts:9:55 - error TS2305: Module '"../../lib"' has no exported member 'KmsKeyProvider'.

    9 import { AppMeshAddOn, EksBlueprint, GlobalResources, KmsKeyProvider } from "../../lib";
```

**Review by shapirov103 — APPROVED** [review]
lgtm

**Pull request merged**